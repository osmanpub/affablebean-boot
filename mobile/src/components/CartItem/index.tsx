import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Button, Image, Text, TextInput, View} from 'react-native';
import {updateProductInCart} from '../../net/cart';

export default class CartItem extends Component {
  constructor(props) {
    super(props);
    const {item} = this.props;

    this.state = {
      qty: item.quantity,
    };
  }

  getProductIcon = name => {
    var icon = '';

    switch (name) {
      case 'broccoli':
        icon = require('../../../assets/img/products/broccoli.png');
        break;

      case 'butter':
        icon = require('../../../assets/img/products/butter.png');
        break;

      case 'cheese':
        icon = require('../../../assets/img/products/cheese.png');
        break;

      case 'chicken leg':
        icon = require('../../../assets/img/products/chicken-leg.png');
        break;

      case 'chocolate cookies':
        icon = require('../../../assets/img/products/chocolate-cookies.png');
        break;

      case 'corn on the cob':
        icon = require('../../../assets/img/products/corn-on-the-cob.png');
        break;

      case 'free range eggs':
        icon = require('../../../assets/img/products/free-range-eggs.png');
        break;

      case 'granola':
        icon = require('../../../assets/img/products/granola.png');
        break;

      case 'green tea':
        icon = require('../../../assets/img/products/green-tea.png');
        break;

      case 'herbal tea':
        icon = require('../../../assets/img/products/herbal-tea.png');
        break;

      case 'jumbo oats':
        icon = require('../../../assets/img/products/jumbo-oats.png');
        break;

      case 'milk':
        icon = require('../../../assets/img/products/milk.png');
        break;

      case 'organic meat patties':
        icon = require('../../../assets/img/products/organic-meat-patties.png');
        break;

      case 'organic coffee':
        icon = require('../../../assets/img/products/organic-coffee.png');
        break;

      case 'parma ham':
        icon = require('../../../assets/img/products/parma-ham.png');
        break;

      case 'porridge oats':
        icon = require('../../../assets/img/products/porridge-oats.png');
        break;

      case 'pumpkin seed bun':
        icon = require('../../../assets/img/products/pumpkin-seed-bun.png');
        break;

      case 'red currants':
        icon = require('../../../assets/img/products/red-currants.png');
        break;

      case 'rice flakes':
        icon = require('../../../assets/img/products/rice-flakes.png');
        break;

      case 'sausages':
        icon = require('../../../assets/img/products/sausages.png');
        break;

      case 'seedless watermelon':
        icon = require('../../../assets/img/products/seedless-watermelon.png');
        break;

      case 'sesame seed bagel':
        icon = require('../../../assets/img/products/sesame-seed-bagel.png');
        break;

      case 'sunflower seed loaf':
        icon = require('../../../assets/img/products/sunflower-seed-loaf.png');
        break;

      case 'wholebean coffee':
        icon = require('../../../assets/img/products/wholebean-coffee.png');
        break;

      default:
    }

    return icon;
  };

  onChange = qty => {
    this.setState({qty});
  };

  updateCart = id => {
    let qty = Number(this.state.qty);

    if (isNaN(qty) || qty < 0 || qty > 10) {
      const {item} = this.props;
      this.setState({qty: item.quantity});
    } else {
      const {dispatch} = this.props;
      dispatch(updateProductInCart(id, qty));
    }
  };

  render() {
    const {item} = this.props;
    const {product} = item;
    const name = product.name;

    return (
      <View
        style={{
          alignItems: 'center',
          flexDirection: 'row',
          paddingBottom: 24,
          paddingLeft: 32,
        }}>
        <Image source={this.getProductIcon(name)} />
        <View style={{paddingLeft: 24}} />
        <View style={{flexDirection: 'column'}}>
          <Text style={{fontWeight: 'bold'}}>{name}</Text>
          <Text style={{fontWeight: 'bold'}}>
            &euro; {item.total.toFixed(2)}
          </Text>
          <Text style={{paddingBottom: 8}}>
            &euro; {product.price.toFixed(2)}
          </Text>
          <TextInput
            onChangeText={qty => this.onChange(qty)}
            placeholder="Enter quantity"
            style={{
              borderBottomColor: 'black',
              borderBottomWidth: 1,
              marginBottom: 8,
            }}
            value={this.state.qty.toString()}
          />
          <Button onPress={() => this.updateCart(product.id)} title="update" />
        </View>
      </View>
    );
  }
}
