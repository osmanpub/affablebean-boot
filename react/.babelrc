 
{
  "presets": ["@babel/preset-react"],
  "env": {
    "test": {
      "plugins": [
        ["istanbul", {
          "exclude": [
            "**/*.spec.js"
          ]
        }]
      ]
    }
  },
  "plugins": ["transform-class-properties", "istanbul"]
}
