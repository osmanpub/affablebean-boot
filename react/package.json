{
  "name": "affablebean",
  "version": "0.1.0",
  "private": true,
  "eslintConfig": {
    "extends": "react-app",
    "plugins": [
      "react-hooks"
    ],
    "rules": {
      "react-hooks/rules-of-hooks": "error",
      "react-hooks/exhaustive-deps": "warn"
    }
  },
  "scripts": {
    "build": "react-scripts build",
    "coveralls": "cat coverage/lcov.info | coveralls",
    "cypress:open": "cypress open",
    "cypress:run": "cypress run",
    "cypress:verify": "cypress verify",
    "eject": "react-scripts eject",
    "report:coverage": "nyc report --reporter=html",
    "report:coverage:text": "nyc report --reporter=text",
    "start": "node_modules/react-scripts/bin/react-scripts.js start &",
    "start:java": "cd .. && sh server && cd react && node_modules/react-scripts/bin/react-scripts.js start &",
    "test": "react-scripts test"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "dependencies": {
    "@reduxjs/toolkit": "^1.2.5",
    "axios": "^0.19.2",
    "node-rest-client": "^3.1.0",
    "react": "^16.12.0",
    "react-alert": "^6.0.0",
    "react-alert-template-basic": "^1.0.0",
    "react-dom": "^16.12.0",
    "react-hook-form": "^4.9.7",
    "react-redux": "^7.2.0",
    "react-router-dom": "^5.1.2",
    "react-scripts": "3.4.0",
    "redux": "^4.0.5",
    "styled-components": "^5.0.1"
  },
  "devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/preset-react": "^7.8.3",
    "@cypress/code-coverage": "^1.12.2",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "^9.4.0",
    "@testing-library/user-event": "^7.2.1",
    "@types/axios": "^0.14.0",
    "@types/cypress": "^1.1.3",
    "@types/jest": "^25.1.3",
    "@types/react": "^16.9.21",
    "@types/react-alert": "^5.2.0",
    "@types/react-bootstrap": "^0.32.20",
    "@types/react-dom": "^16.9.5",
    "@types/react-redux": "^7.1.7",
    "@types/react-router-dom": "^5.1.3",
    "@types/react-test-renderer": "^16.9.2",
    "@types/redux": "^3.6.0",
    "@types/redux-mock-store": "^1.0.2",
    "@types/styled-components": "^5.0.0",
    "babel-plugin-istanbul": "^6.0.0",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "coveralls": "^3.0.9",
    "cross-env": "^7.0.0",
    "cypress": "^4.0.2",
    "istanbul-lib-coverage": "^3.0.0",
    "nyc": "^15.0.0",
    "react-test-renderer": "^16.12.0",
    "redux-mock-store": "^1.5.4",
    "typescript": "^3.8.2"
  },
  "nyc": {
    "sourceMap": false,
    "instrument": false
  }
}
