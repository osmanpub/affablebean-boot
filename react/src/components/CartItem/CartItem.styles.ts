import styled from "styled-components";

export const CartTableTd = styled.td`
  width: 145px;
  height: 90px;
`;
