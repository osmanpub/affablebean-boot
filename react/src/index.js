import React from "react";
import { render } from "react-dom";
import { configureStore } from "redux-starter-kit";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route } from "react-router-dom";
import reducer from "./reducers";
import App from "./containers/App";
import CategoryProducts from "./containers/CategoryProducts";
import * as serviceWorker from "./serviceWorker";

const store = configureStore({
  reducer
});

render(
  // <React.StrictMode>
  <Provider store={store}>
    <Router>
      <Route path="/" exact component={App} />
      <Route path="/category/:id" component={CategoryProducts} />
    </Router>
  </Provider>,
  // </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
