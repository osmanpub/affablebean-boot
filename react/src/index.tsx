import { configureStore } from "@reduxjs/toolkit";
import React, { lazy, Suspense } from "react";
import { render } from "react-dom";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import ErrorBoundary from "./components/ErrorBoundary";
import rootReducer from "./redux";
import * as serviceWorker from "./serviceWorker";
const Cart = lazy(() => import("./containers/Cart"));
const CategoryProducts = lazy(() => import("./containers/CategoryProducts"));
const Checkout = lazy(() => import("./containers/Checkout"));
const Contact = lazy(() => import("./containers/Contact"));
const Home = lazy(() => import("./containers/Home"));
const Privacy = lazy(() => import("./containers/Privacy"));

const store = configureStore({
  reducer: rootReducer
});

render(
  <React.StrictMode>
    <ErrorBoundary>
      <Provider store={store}>
        <Router>
          <Suspense fallback={<div>Loading...</div>}>
            <Switch>
              <Route path="/" exact component={Home} />
              <Route path="/category/:id" component={CategoryProducts} />
              <Route path="/checkout" component={Checkout} />
              <Route path="/contact" component={Contact} />
              <Route path="/privacy" component={Privacy} />
              <Route path="/viewCart/:clear" component={Cart} />
            </Switch>
          </Suspense>
        </Router>
      </Provider>
    </ErrorBoundary>
  </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
