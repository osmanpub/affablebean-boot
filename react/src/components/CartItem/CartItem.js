import React, { Component } from "react";
import { updateProductInCart } from "../../net/cart";
import { CartTableTd } from "./CartItem.styles";

export class CartItem extends Component {
  constructor(props) {
    super(props);
    const { item } = this.props;

    this.state = {
      qty: item.quantity
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const inputQty = event.target;
    let qty = Number(inputQty.value);

    if (qty >= 0 && qty <= 10) {
      this.setState({ qty: qty });
    } else {
      inputQty.value = this.state.qty;
    }
  }

  handleSubmit(event) {
    event.preventDefault();
    const { dispatch, item } = this.props;
    const { product } = item;
    dispatch(updateProductInCart(product.id, Number(this.state.qty)));
  }

  render() {
    const { index, item } = this.props;
    const { product } = item;
    const name = product.name;
    const rowCol = index % 2 === 0 ? "white" : "lightBlue";
    const inputStyle = {
      margin: "5px",
      textAlign: "center"
    };

    return (
      <React.Fragment>
        <tr className={`${rowCol}`}>
          <CartTableTd>
            <img src={`/static/img/products/${name}.png`} alt="{name}" />
          </CartTableTd>
          <CartTableTd>{name}</CartTableTd>
          <CartTableTd>
            &euro;
            {item.total.toFixed(2)}
            <br />
            {product.price}
          </CartTableTd>
          <CartTableTd>
            <form onSubmit={this.handleSubmit}>
              <input
                className="form-control"
                maxLength="2"
                onChange={this.handleChange}
                size="2"
                style={inputStyle}
                type="number"
                value={this.state.qty}
              />
              <button className="`btn btn-primary btn-sm`" type="submit">
                update
              </button>
            </form>
          </CartTableTd>
        </tr>
      </React.Fragment>
    );
  }
}
