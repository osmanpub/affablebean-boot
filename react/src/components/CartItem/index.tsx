import React, { ChangeEvent, useState } from "react";
import { useDispatch } from "react-redux";
import { getId } from "../../helpers/utils";
import { CartItem as CartItemState } from "../../interfaces/cart";
import { updateProductInCart } from "../../net/cart";
import { CartTableTd } from "./CartItem.styles";

type Props = {
  index: number;
  item: CartItemState;
};

export default function CartItem(props: Props) {
  const { index, item } = props;
  const { product } = item;
  const [state, setState] = useState({ qty: item.quantity });
  const dispatch = useDispatch();

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const input = event.target;
    let qty = Number(input.value);

    if (qty >= 0 && qty <= 10) {
      setState({ ...state, qty });
    } else {
      input.value = qty.toString();
    }
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    dispatch(updateProductInCart(getId(product), Number(state.qty)));
  };

  const name = product.name;
  const rowCol = index % 2 === 0 ? "white" : "lightBlue";

  return (
    <React.Fragment>
      <tr className={`${rowCol}`} data-cy={`product-${name}`}>
        <CartTableTd>
          <img src={`/static/img/products/${name}.png`} alt="{name}" />
        </CartTableTd>
        <CartTableTd>
          <span>{name}</span>
        </CartTableTd>
        <CartTableTd>
          &euro;
          {item.total.toFixed(2)}
          <br />
          {product.price}
        </CartTableTd>
        <CartTableTd>
          <form onSubmit={handleSubmit}>
            <input
              data-cy={`input-qty-${name}`}
              className="form-control"
              maxLength={2}
              onChange={handleChange}
              size={2}
              style={{ margin: "5px", textAlign: "center", width: "90%" }}
              type="number"
              value={state.qty}
            />
            <button
              className="btn btn-primary btn-sm"
              type="submit"
              data-cy={`update-qty-${name}`}
            >
              update
            </button>
          </form>
        </CartTableTd>
      </tr>
    </React.Fragment>
  );
}
