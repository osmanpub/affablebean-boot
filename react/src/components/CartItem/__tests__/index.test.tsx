import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, waitForElement } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import CartItem from "..";

const mockStore = configureStore([]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [],
    numberOfItems: 0,
    subtotal: 0
  }
};

const store = mockStore(initialState);

const item = {
  product: {
    id: 1,
    description: "semi skimmed (1L)",
    name: "milk",
    price: 1.7,
    category: { id: 1, name: "dairy" },
    _links: {
      self: { href: "http://localhost:8080/api/products/1" },
      products: { href: "http://localhost:8080/api/products" }
    }
  },
  quantity: 1,
  total: 1.7
};

const cartItem = (
  <Provider store={store}>
    <CartItem index={0} item={item} />
  </Provider>
);

describe("<CartItem />", () => {
  it("renders correctly", () => {
    const component = renderer.create(cartItem).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("renders correctly", () => {
    const { getByText, getByRole, container, asFragment } = render(cartItem);
    const updateButton = getByText(/update/i);
    expect(updateButton).toBeInTheDocument();
  });
});
