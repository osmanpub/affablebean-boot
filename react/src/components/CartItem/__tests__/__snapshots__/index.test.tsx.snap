// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> renders correctly 1`] = `
<tr
  className="white"
  data-cy="product-milk"
>
  <td
    className="sc-AxjAm jhCEof"
  >
    <img
      alt="milk"
      src="/static/img/products/milk.png"
    />
  </td>
  <td
    className="sc-AxjAm jhCEof"
    data-testid="milk"
  >
    milk
  </td>
  <td
    className="sc-AxjAm jhCEof"
    data-testid={1.7}
  >
    â‚¬
    1.70
    <br />
    1.7
  </td>
  <td
    className="sc-AxjAm jhCEof"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        className="form-control"
        data-cy="input-qty-milk"
        data-testid="input-qty-milk"
        defaultValue={1}
        name="quantity"
        size={2}
        style={
          Object {
            "margin": "5px",
            "textAlign": "center",
            "width": "90%",
          }
        }
        type="number"
      />
      <button
        className="btn btn-primary btn-sm"
        data-cy="update-qty-milk"
        data-testid="update-qty-milk"
        type="submit"
      >
        update
      </button>
    </form>
  </td>
</tr>
`;
